### PR TITLE
fix: Add functionality for creating recurring events

### LINKED ISSUE #1583: Fix Recurring Events
---
### Note

This is a very important feature to implement. 
1. We need someone with either demonstrable:
    1. experience with GraphQL
    2. activity in their GitHub account with a focus on TypeScript
1. This is not an issue for a first time attempt at the technology. 
    1. The first PR needs to be created within a week
    1. All PRs must be completed by the end of February
1. Read through the comments in the closed PR about the progress and why the approach was selected. There are further details below.

In summary:
1. This is not for a newbie
1. This is not for someone with limited time over the next month
---

**Describe the bug**
The creation of a recurring event only creates a single event in the API clients.

**To Reproduce**

1. Create a recurring event
3. Only a single instance is visible

**Expected behavior**

1. You'll need to update or create new mutations that include start and end date for the event recurrence, not just the event itself
1. There will need to be a query created or updated that would return all the calendar entries for a particular time frame including the recurring ones.
1. We will need the ability to edit any recurrence of the event at any time that the event occurs. The edits would need to be valid for:
    1. All instances in the series
    2. Just the instance in question
    4. The instance in question and all those in future
1. Administrators will need to be able to track attendance for any type of event. Therefore you will need to think of ways to create permanence for historical recurring events for this purpose
1. We don't want the creation of an infinite number of event calendar entries if a recurring event has no end date
1. Other functionality must not be affected
1. You must write tests to validate your work

**Actual behavior**
- See above

**Screenshots**
- N/A

**Additional details**
1. Work with @xoldyckk on evaluating your solution here and in the subsequent PR
1. This is a prerequisite for our menu of project ideas for GSoC 2024
1.  When this issue is resolved we'll need to create another one to support custom recurrences, such as `3rd week of the month`, similar to what is available in Google Calendar
1. We almost completed the work for Weekly recurring when the assignee @Community-Programmer was unable to continue due to exam time constraints. 
    1. The first PR for this issue will need to build on the work in this abandoned PR as it was almost approved for merging. 
        1. https://github.com/PalisadoesFoundation/talawa-api/pull/1738
    1. You will need to submit a plan as described here
         1. https://github.com/PalisadoesFoundation/talawa-api/pull/1738#issuecomment-1910724294
1. You will need to split this issue into multiple PRs, each with custom options provided by the RRULE package
    1. Weekly recurring
    2. Daily recurring
    3. Monthly recurring
    5. Annually recurring

**Potential internship candidates**
- Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by meetulr**
**What kind of change does this PR introduce?**
It adds the functionality for creating recurring events based on their recurrence patterns.

**Issue Number:**

Fixes #1583

**Did you add tests for your changes?**

Yes

**Does this PR introduces any breaking changes?**

No

**Summary**

To implement creation and management of recurring events, we're following [this](https://meetulr.github.io/Recurrence-Plan/) approach.
This PR implements the first part of it. i.e. creation of recurring events.

**Additional Context**

This is the first in the series of PRs for this issue, after this we would be adding the functionalities for querying, updating and deleting recurring events.

### CODE REVIEW

**Comment by meetulr** [comment]
No changes were made on the frontend regarding this, so by default it creates a weekly recurring event:

https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/f9105177-1b55-4fd6-962e-b2f04263ae54


We could create custom recurrence patterns:

https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/5c7cebbe-5b1c-4f5b-ab17-fbb2d0c11401


https://github.com/PalisadoesFoundation/talawa-api/assets/55585268/124871cd-bcf1-4e1d-a571-f9439044cfe1

**Comment by meetulr** [comment]
@palisadoes The number of files are large because I moved helper functions into their separate files, for ease of understanding and use later on. And also removed some of the files that were merged with previous PRs for this issue.
This PR is about creating recurring events. I've structured the code this way to make things easier for further steps (query, updation, deletion).
So, subsequent PRs won't contain these many files. This PR sort of sets things up for future work.

**Comment by Cioppolo14** [comment]
@meetulr Please fix the failing tests.

**Comment by meetulr** [comment]
The test if failing because there are formatting issues in many files (files not associated with this PR), and it's running the check on all of them. Due to this other tests are skipped too. I think it's happening since the testing behaviour changed #1799. It's running the test on all the files with ".ts" extension.
The files in this PR are already linted by the pre-commit hook.

**Comment by palisadoes** [comment]
Please do a minor commit. I made a change that hopefully should fix it.

**Comment by meetulr** [comment]
This is ready for review.

File: src/helpers/event/createEventHelpers/createRecurringEvent.ts [hunk 1]
```diff
@@ -49,20 +48,14 @@ export const createRecurringEvent = async (
...
   // create a base recurring event first, based on which all the
   // recurring instances would be dynamically generated
   const baseRecurringEvent = await Event.create(
     [
       {
         ...data,
-        startDate: convertedStartDate,
-        endDate: convertedEndDate,
+        startDate: data.startDate,
+        endDate: data.endDate,
```
Comment (SiddheshKukade) [inline src/helpers/event/createEventHelpers/createRecurringEvent.ts]: are you assuming that sender would send the date in UTC format ? or should we perform the convertToUTCDate() here?

File: src/helpers/event/recurringEventHelpers/generateRecurringEventInstances.ts [hunk 2]
```diff
@@ -51,12 +50,10 @@ export const generateRecurringEventInstances = async ({
...
   const recurringInstances: InterfaceRecurringEvent[] = [];
   recurringInstanceDates.map((date) => {
-    const formattedInstanceDate = format(date, "yyyy-MM-dd");
-
     const createdEventInstance = {
       ...data,
-      startDate: formattedInstanceDate,
-      endDate: formattedInstanceDate,
+      startDate: date,
+      endDate: date,
```
Comment (SiddheshKukade) [inline src/helpers/event/recurringEventHelpers/generateRecurringEventInstances.ts]: Also make sure abut date formatting here

**Review by SiddheshKukade — CHANGES_REQUESTED** [review]

File: src/helpers/event/createEventHelpers/createRecurringEvent.ts (hunk 1 above)
Comment (meetulr) [reply src/helpers/event/createEventHelpers/createRecurringEvent.ts]: Yes, the dates are formatted by already by dayjs, we're recieving dates in UTC. That's why I removed the unnecessary date formatting. I've been testing this for past couple of days.

File: src/helpers/event/recurringEventHelpers/generateRecurringEventInstances.ts (hunk 2 above)
Comment (meetulr) [reply src/helpers/event/recurringEventHelpers/generateRecurringEventInstances.ts]: same here, rrule returns the dates in proper format, no need to format it again.

**Comment by SiddheshKukade** [comment]
@meetulr can you provide answer to my questions so that I can review again ?

**Comment by meetulr** [comment]
> @meetulr can you provide answer to my questions so that I can review again ?

I did in the comments. I'm sorry I shouldn't have resolved it so soon😅.

> are you assuming that sender would send the date in UTC format ? or should we perform the convertToUTCDate() here?

Yes, the dates are already formatted by dayjs on the frontend, and by graphQL, we're receiving dates in UTC midnight in the backend. I didn't realize that before adding the convertToUTCDate and timeZoneOffset utils, and formatting dates using date-fns's format... causing unnecessary confusion.
That wasn't required, so I removed it later. And that's why I've converted the dates to UTC midnight in the tests to simulate this behavior.
I've been testing this for past couple of days.

> Also make sure abut date formatting here

Same here, rrule returns the dates in proper format, no need to format it again. It would only serve to cause timezone discrepencies and incorrect date output.

**Review by EshaanAgg — APPROVED** [review]
LGTM. Would just advice you to add the documentation about all of the logic of this recurrence somewhere. Also you need to now create new events based on the queries that come in, so please take a look into that as well.

**Comment by meetulr** [comment]
> LGTM. Would just advice you to add the documentation about all of the logic of this recurrence somewhere. Also you need to now create new events based on the queries that come in, so please take a look into that as well.

Yes, I have added docs and comments in the files. Would add some more if required.

The next PR (for query i.e. dynamic generation) is almost ready. Just have to write the tests.

**Review by palisadoes — APPROVED** [review]

**Pull request merged**